### PR TITLE
Task4

### PR DESCRIPTION
**Pull request by alexkarlov**
I didn't find benefits of usage of usage Async storage for passing a context. I think using a context inside a request is much more common and convinient way, IMHO

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -2,13 +2,35 @@ const express = require("express");
...
 app.post("/events", handlers.createEvent);
 app.put("/events/:eventId", handlers.updateEvent);
 app.delete("/events/:eventId", handlers.deleteEvent);
 app.get("/events-batch", handlers.getEventsBatch);
-app.listen(3000, function () {
-  console.log("listening on 3000");
-});
+app.use(handlers.errorHandler);
+storage.run(
+  {
```
Comment (viktornord) [inline app.js]: did you try to get any content in the middlewares ? we should run out handlers within run phase

File: logger.js [hunk 2]
```diff
@@ -1,25 +1,58 @@
 class Logger {
```
Comment (viktornord) [inline logger.js:3]: hw required to have a lib integrate: e.g. pino, winston

File: app.js [hunk 3]
```diff
@@ -2,13 +2,35 @@ const express = require("express");
...
+    traceID: uuid.v4(),
+  };
+  req.context = ctx;
+  Logger.log(req, "new request");
+  next();
+});
 app.use(express.json());
-app.get("/events/:eventId", handlers.getEventById);
+app.get("/events/:eventId", (req, res) => {
+  // i don't think this is a good solution to pass storage as an additional parameter, but...
```
Comment (viktornord) [inline app.js]: no need, why can't you have it separated and required in handlers?

**Review by viktornord — CHANGES_REQUESTED** [review]

**Comment by alexkarlov** [comment]
Sorry, forgot to push the latest changes, now it should look better

File: app.js [hunk 4]
```diff
@@ -2,13 +2,39 @@ const express = require("express");
...
+  };
+  req.context = ctx;
+  Logger.log(req, "new request");
+  next();
+});
 app.use(express.json());
-app.get("/events/:eventId", handlers.getEventById);
+app.get("/events/:eventId", (req, res) => {
+  const id = storage.getStore().get("requestId");
+  console.log(`[${id}] request received`);
```
Comment (viktornord) [inline app.js]: why not to use logger?

File: app.js [hunk 5]
```diff
@@ -2,13 +2,39 @@ const express = require("express");
...
+const requestIdMiddleware = (req, res, next) => {
+  storage.run(new Map(), () => {
+    storage.getStore().set("requestId", uuid.v4());
+    next();
+  });
+};
+
+app.use(requestIdMiddleware);
+app.use(function (req, res, next) {
+  ctx = {
```
Comment (viktornord) [inline app.js:18]: why do we need that? if we have requestId?

File: app.js [hunk 6]
```diff
@@ -2,13 +2,39 @@ const express = require("express");
...
 const path = require("path");
 const handlers = require(path.join(__dirname, "events", "handlers.js"));
+const uuid = require("uuid");
+const { AsyncLocalStorage } = require("async_hooks");
+const storage = new AsyncLocalStorage();
+const Logger = require("./logger").createLogger();
+
+const requestIdMiddleware = (req, res, next) => {
+  storage.run(new Map(), () => {
+    storage.getStore().set("requestId", uuid.v4());
```
Comment (viktornord) [inline app.js]: why not to create map with requestId already?

File: events/handlers.js [hunk 7]
```diff
@@ -174,11 +173,17 @@ function lockEvents(req, res, callback) {
   });
 }
 
+function errorHandler(err, req, res, next) {
+  console.error(err.stack);
```
Comment (viktornord) [inline events/handlers.js]: why not to use logger?

**Review by viktornord — CHANGES_REQUESTED** [review]

File: app.js [hunk 8]
```diff
@@ -2,13 +2,38 @@ const express = require("express");
...
+    next();
+  });
+};
+
+app.use(requestIdMiddleware);
+app.use(function (req, res, next) {
+  ctx = {
+    traceID: storage.getStore().get("requestId"),
+  };
+  req.context = ctx;
```
Comment (viktornord) [inline app.js:21]: why do we need to put context in request if we can grab it from storage almost anywhere

File: logger.js [hunk 9]
```diff
@@ -1,25 +1,69 @@
...
+  error(...msg) {
+    const extra = extractReq(msg);
+    if (extra !== null) {
+      msg[0] = extra;
     }
+    this.logger.error(...msg);
+  }
+
+  warn(...msg) {
+    const extra = extractReq(msg);
```
Comment (viktornord) [inline logger.js:38]: all this can be put in storage

**Review by viktornord — APPROVED** [review]

File: app.js (hunk 8 above)
Comment (alexkarlov) [reply app.js:21]: It's a question of habit. I'd prefer to keep the context near (inside) the request

**Pull request merged**